src/server/scripts/Outland/TempestKeep/Eye/boss_kaelthas.cpp

```diff
@@ -456,557 +438,290 @@ class boss_kaelthas : public CreatureScript
-                                }
-
-                                MindControl_Timer = 60000;
+                                if (Unit* target = SelectTarget(SELECT_TARGET_RANDOM, 3))
+                                    DoCast(target, SPELL_MIND_CONTROL);
                             }
-                            else
-                                MindControl_Timer -= diff;
-                        }
-
-                        //Phoenix_Timer
-                        if (Phoenix_Timer <= diff)
-                        {
+                            events.ScheduleEvent(EVENT_MIND_CONTROL, 60000);
+                            break;
+                        case EVENT_SUMMON_PHOENIX:
                             DoCast(me, SPELL_PHOENIX_ANIMATION);
                             Talk(SAY_SUMMON_PHOENIX);
+                            events.ScheduleEvent(EVENT_SUMMON_PHOENIX, 60000);
+                            break;
+                        case EVENT_SHOCK_BARRIER:
+                            DoCast(me, SPELL_SHOCK_BARRIER);
 
-                            Phoenix_Timer = 60000;
-                        }
-                        else
-                            Phoenix_Timer -= diff;
-
-                        //Phase 4 specific spells
-                        if (Phase == 4)
-                        {
-                            if (HealthBelowPct(50))
-                            {
-                                instance->SetData(DATA_KAELTHASEVENT, 4);
-                                Phase = 5;
-                                Phase_Timer = 10000;
-
-                                Talk(SAY_PHASE5_NUTS);
-
-                                me->StopMoving();
-                                me->GetMotionMaster()->Clear();
-                                me->GetMotionMaster()->MoveIdle();
-                                me->SetPosition(afGravityPos[0], afGravityPos[1], afGravityPos[2], 0);
-                                me->MonsterMoveWithSpeed(afGravityPos[0], afGravityPos[1], afGravityPos[2], 1);
-
-                                me->InterruptNonMeleeSpells(false);
-                                DoCast(me, SPELL_FULLPOWER);
-                                me->SetFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_NON_ATTACKABLE);
-                            }
-
-                            //ShockBarrier_Timer
-                            if (ShockBarrier_Timer <= diff)
+                            if (events.IsInPhase(PHASE_TRANSITION) || _inGravityLapse)
+                                events.ScheduleEvent(EVENT_SHOCK_BARRIER, _inGravityLapse ? 60000 : 20000);
+                            break;
+                        case EVENT_END_TRANSITION:
+                            DoCast(me, SPELL_EXPLODE);
+                            me->InterruptNonMeleeSpells(false);
+                            me->RemoveAurasDueToSpell(SPELL_FULLPOWER);
+                            me->RemoveFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_NON_ATTACKABLE);
```

Comment:
I ended up opening the PR with an older version of the file but anyway I know that it should remove the not selectable flag in this line instead.